src/guest_memory.rs

```diff
@@ -162,6 +163,15 @@ pub trait GuestMemoryRegion: Bytes<MemoryRegionAddress, E = Error> {
             .and_then(|offset| self.check_address(MemoryRegionAddress(offset)))
     }
 
+    /// Return the file descriptor pointing to the memory region. Return None
+    /// if no file descriptor is associated with the region.
+    fn fd(&self) -> Option<RawFd>;
```

Comment:
> Also, just wondering, what's the use case for having them as part of the `GuestMemoryRegion` trait? If none of the `vm-memory` consumers actually needs something like this, I would remove them for now.

Needed for vhost.

Reply:
>Since fd and fd_offset are either both Some(...) or both None, I think it would be nicer to have a single method that returns an Option<(RawFd, usize)>.

Yes I agree, I'll change that.

>Also, just wondering, what's the use case for having them as part of the GuestMemoryRegion trait? If none of the vm-memory consumers actually needs something like this, I would remove them for now.

Here is a consumer https://github.com/intel/cloud-hypervisor/pull/30 ;)

>Btw, I think RawFd doesn't work for Windows; we'll either have to abstract this (like provide a reference to the actual File or smt), or conditionally compile different versions of the method for different platforms.

Ah yes, what's the preferred way to do this? @sameo @rbradford @jiangliu

Reply:
Well, one way of not being biased towards `unix` stuff in this particular case is to return an `Option<&File>` instead of `Option<RawFd>` (assuming you have a `File` object somewhere, which seems to be the case).

Reply:
Both Raw and File should be OK, but not AsRawFd because it's not Send ready. We could easily construct a File object from RawFd and File is Send/Sync.

Reply:
Well we would not have to construct the File if we were actually keeping the File. WDYT?

Reply:
Yeah, I mean File is as convenient as RawFd, and it's more friend to Windows:)

Reply:
Hi, I've created a PR [here](https://github.com/sboeuf/vm-memory/pull/2) with some changes on top of Sebastien's branch that embody the proposals I've made. I don't know where those conflicts came from :( Still, it would be great if you ppl can have a look.

Reply:
File is better not really because of Windows, since this is meant for SCM_RIGHTS support, but rather because it allows tracking the lifetime correctly. What about a separate struct FileRange:

```
struct FileRange {
    file: Arc<File>,
    offset: u64,
    size: usize
};
```

and having a single method that returns `Option<FileRange>` here?